src/Psalm/Type/Atomic/TKeyOfArray.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Psalm\Type\Atomic;
+
+use Psalm\Type\Atomic;
+
+/**
+ * Represents an offset of an array.
+ *
+ * @psalm-type ArrayLikeTemplateType = TClassConstant|TKeyedArray|TList|TArray
+ */
+class TKeyOfArray extends TArrayKey
+{
+    /** @var ArrayLikeTemplateType */
+    public $type;
```

Comment:
So, If I understand this right, whenever you need to create a TKeyOfArray, you store the type of the array at the moment of the TKeyOfArray creation.

Does this mean that if the array is refined after the creation of the TKeyOfArray (for example, with isset on specific keys), the TKeyOfArray will have obsolete information?

Also, Is there a reason to store the entire array? Wouldn't the key types be enough?

Reply:
> Does this mean that if the array is refined after the creation of the TKeyOfArray (for example, with isset on specific keys), the TKeyOfArray will have obsolete information?
I'm not sure how that would be possible, I now save the reference to the union to support unions. Does this union get updated on asserts?

Reply:
Don't mind the first part, I didn't grasp the change well enough when I wrote that.

About the second part, I'd have said `getGenericArrayType` for TKeyedArray, TInt for TList and type_params[0] for TArray, but it may create issues, I'm not sure

Reply:
I'm also not sure about it, and for constants the key-of/value-of implementation would still need to be duplicated inside `TypeExpander`.
At least I think it's a bit easier to understand the code like this.

Reply:
Nevermind then :)